config.yml

```diff
@@ -1,98 +1,121 @@
-  secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_projects_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      delete: false  # Set to true to delete this variable
-
-  # keys and values here are as described at https://docs.gitlab.com/ee/api/group_level_variables.html#create-variable
-  group_secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_groups_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      variable_type: env_var #or file
-      protected: false #
-      delete: false  # Set to true to delete this variable
-
-group_settings:
-  # settings for ALL projects in 'my-group' group
-  "my-group":
-    deploy_keys:
-      # this name is not actually used, it's just for you
-      a_friendly_deploy_key_name:
-        # you have to pass whole SSH key content here even if GitLab already has this key added and you just assign it
-        # to another project here
-        # this is a limitation of GitLab API
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDB2QKx6BPzL (...)"
-        title: ssh_key_name_that_is_shown_in_gitlab
-        # note that you can set this to `true` or `false` only on the first assignment of the key / its creation
-        # see https://gitlab.com/gitlab-org/gitlab-ce/issues/30021#note_39567845
-        # this is a limitation of GitLab API
-        can_push: false
-      # this name is not actually used, it's just for you
-      another_friendly_deploy_key_name:
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDtbyEK66RXg (...)"
-        title: other_ssh_key_name_that_is_shown_in_gitlab
-        can_push: false
 
-    # keys and values here are as described at https://docs.gitlab.com/ee/api/project_level_variables.html#create-variable
-    secret_variables:
-      # this name is not actually used, it's just for you
-      a_friendly_secret_variable_name:
-        key: SSH_PRIVATE_KEY_BASE64
-        value: "LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUl (...)"
+# The main config starts with this key
+projects_and_groups:
+
+  # On this indentation level you choose what groups and/or projects you want to configure.
+  #
+  # Here are examples of possible dict key.
+  #
+  # This matches all the groups AND all the projects EXCEPT personal projects!
+  # "*":
+  #
+  # This matches all the personal projects.
+  # "<all-users>":
+  #
+  # This matches the "my-group" group AND projects in it.
+  # my-group:
+  #
+  # This matches the "my-subgroup" subgroup within the "my-group" group AND projects in it.
+  # my-group/my-subgroup:
+  #
+  # This matches all the groups that have names matching the '(foo)+' regexp AND projects in it.
+  # "~(foo)+":
+  #
+  # This matches all the subgroups of group "my-group" that have names matching the '(bar)+' regexp
```

Comment:
> subgroups of group "my-group" that have names matching the '(bar)+' regexp  AND projects in it.

I'm guessing **projects** directly under `my-group` with matching name would be included here as well?

And if not, how could one express it?

Reply:
Another good point that makes me think we should really consider going toward recursive search...